[CI] QueryableReservedRolesIT testConfiguredReservedRolesAfterClosingAndOpeningIndex failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #12207 / windows-2022_checkpart4_platform-support-windows](https://gradle-enterprise.elastic.co/s/uf6a252dboum6)
- [elasticsearch-periodic-platform-support #12207 / windows-2022_checkpart4_platform-support-windows](https://gradle-enterprise.elastic.co/s/rtp6qqfuvm5tu)
- [elasticsearch-periodic-platform-support #12185 / windows-2019_checkpart4_platform-support-windows](https://gradle-enterprise.elastic.co/s/uqm42vx27dwwi)
- [elasticsearch-periodic-platform-support #12180 / windows-2022_checkpart4_platform-support-windows](https://gradle-enterprise.elastic.co/s/3qtqgelmfgxja)

**Reproduction Line:**
```
gradlew ":x-pack:plugin:security:qa:security-basic:javaRestTest" --tests "org.elasticsearch.xpack.security.QueryableReservedRolesIT.testConfiguredReservedRolesAfterClosingAndOpeningIndex" -Dtests.seed=B50EB6F18367D92 -Dtests.locale=ar-MR -Dtests.timezone=CNT -Druntime.java=26
```

**Applicable branches:**
8.19

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.security.QueryableReservedRolesIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testConfiguredReservedRolesAfterClosingAndOpeningIndex'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: <[apm_system, editor, viewer, beats_system, logstash_system, kibana_system, superuser]>
     but: was <[viewer, editor, superuser]>
```

**Issue Reasons:**
- [8.19] 2 consecutive failures in step windows-2022_checkpart4_platform-support-windows
- [8.19] 4 failures in test testConfiguredReservedRolesAfterClosingAndOpeningIndex (1.7% fail rate in 234 executions)
- [8.19] 3 failures in step windows-2022_checkpart4_platform-support-windows (37.5% fail rate in 8 executions)
- [8.19] 3 failures in pipeline elasticsearch-periodic-platform-support (50.0% fail rate in 6 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -440,9 +440,6 @@ tests:
 - class: org.elasticsearch.xpack.sql.qa.single_node.RestSqlIT
   method: testNextPageText
   issue: https://github.com/elastic/elasticsearch/issues/147950
-- class: org.elasticsearch.xpack.security.QueryableReservedRolesIT
-  method: testConfiguredReservedRolesAfterClosingAndOpeningIndex
-  issue: https://github.com/elastic/elasticsearch/issues/147972
 - class: org.elasticsearch.xpack.stateless.StatelessHollowIndexShardsIT
   method: testDataStreamStatsActionBroadcastedToSearchNodesOnly
   issue: https://github.com/elastic/elasticsearch/issues/147973

--- a/x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryableReservedRolesIT.java
+++ b/x-pack/plugin/security/qa/security-basic/src/javaRestTest/java/org/elasticsearch/xpack/security/QueryableReservedRolesIT.java
@@ -295,6 +295,15 @@ public class QueryableReservedRolesIT extends ESRestTestCase {
         );
         Response response = adminClient().performRequest(request);
         assertOK(response);
+        // The close API can return acknowledged with the close-block applied but before IndexMetadata.state has
+        // been flipped to CLOSE.
+        assertBusy(() -> {
+            final Request stateRequest = new Request("GET", "_cluster/state/metadata/" + INTERNAL_SECURITY_MAIN_INDEX_7);
+            final Response stateResponse = adminClient().performRequest(stateRequest);
+            assertOK(stateResponse);
+            final String indexState = ObjectPath.createFromResponse(stateResponse).evaluate("metadata.indices.\\.security-7.state");
+            assertThat(indexState, equalTo("close"));
+        }, 30, TimeUnit.SECONDS);
     }
 
     private void openSecurityIndex() throws Exception {

--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java
@@ -22,6 +22,7 @@ import org.elasticsearch.cluster.NotMasterException;
 import org.elasticsearch.cluster.SimpleBatchedExecutor;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
 import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.MetadataIndexStateService;
 import org.elasticsearch.cluster.metadata.ProjectMetadata;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.cluster.service.MasterServiceTaskQueue;
@@ -487,7 +488,24 @@ public final class QueryableBuiltInRolesSynchronizer implements ClusterStateList
 
     private boolean isSecurityIndexClosed(final ClusterState state) {
         final IndexMetadata indexMetadata = resolveSecurityIndexMetadata(state.metadata());
-        return indexMetadata != null && indexMetadata.getState() == IndexMetadata.State.CLOSE;
+        return indexMetadata != null && isClosedOrClosing(state, indexMetadata);
+    }
+
+    /**
+     * Returns true if the given index is closed or is in the process of being closed. The close API installs
+     * {@link MetadataIndexStateService#INDEX_CLOSED_BLOCK} before flipping {@link IndexMetadata#getState()} to
+     * {@link IndexMetadata.State#CLOSE}, so checking the state field alone misses the in-progress window.
+     */
+    private static boolean isClosedOrClosing(final ClusterState state, final IndexMetadata indexMetadata) {
+        if (indexMetadata.getState() == IndexMetadata.State.CLOSE) {
+            return true;
+        }
+        return state.blocks()
+            .hasIndexBlockWithId(
+                state.metadata().projectFor(indexMetadata.getIndex()).id(),
+                indexMetadata.getIndex().getName(),
+                MetadataIndexStateService.INDEX_CLOSED_BLOCK_ID
+            );
     }
 
     /**
@@ -504,12 +522,13 @@ public final class QueryableBuiltInRolesSynchronizer implements ClusterStateList
         final Map<String, String> newRolesDigests,
         final ActionListener<Void> listener
     ) {
-        final IndexMetadata securityIndexMetadata = resolveSecurityIndexMetadata(clusterService.state().metadata());
+        final ClusterState currentState = clusterService.state();
+        final IndexMetadata securityIndexMetadata = resolveSecurityIndexMetadata(currentState.metadata());
         if (securityIndexMetadata == null) {
             listener.onFailure(new IndexNotFoundException(SECURITY_MAIN_ALIAS));
             return;
         }
-        if (securityIndexMetadata.getState() == IndexMetadata.State.CLOSE) {
+        if (isClosedOrClosing(currentState, securityIndexMetadata)) {
             listener.onFailure(new IndexClosedException(securityIndexMetadata.getIndex()));
             return;
         }
@@ -580,11 +599,9 @@ public final class QueryableBuiltInRolesSynchronizer implements ClusterStateList
             if (indexMetadata == null) {
                 throw new IndexNotFoundException(concreteSecurityIndexName);
             }
-            if (indexMetadata.getState() == IndexMetadata.State.CLOSE) {
-                // Index was closed between when the sync started and when this task runs. Do not update the
-                // synced-roles digest in metadata, since the actual role documents could not have been
-                // updated/deleted in a closed index. IndexClosedException is treated as an expected failure
-                // and does not count toward failedSyncAttempts.
+            if (isClosedOrClosing(state, indexMetadata)) {
+                // Bail if the index closed (or began closing) since the sync started. IndexClosedException is
+                // in isExpectedFailure, so this does not count against failedSyncAttempts.
                 throw new IndexClosedException(indexMetadata.getIndex());
             }
             Map<String, String> existingRoleDigests = indexMetadata.getCustomData(METADATA_QUERYABLE_BUILT_IN_ROLES_DIGEST_KEY);

--- a/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizerTests.java
+++ b/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizerTests.java
@@ -18,6 +18,7 @@ import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.block.ClusterBlocks;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
 import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.MetadataIndexStateService;
 import org.elasticsearch.cluster.metadata.ProjectId;
 import org.elasticsearch.cluster.node.DiscoveryNodeRole;
 import org.elasticsearch.cluster.node.DiscoveryNodeUtils;
@@ -406,6 +407,29 @@ public class QueryableBuiltInRolesSynchronizerTests extends ESTestCase {
         // could not have been updated.
         final ClusterState clusterState = markShardsAvailable(createClusterStateWithClosedSecurityIndex()).nodes(localNodeMaster())
             .blocks(emptyClusterBlocks())
+            .build();
+        final Map<String, String> newRoleDigests = Map.of("superuser", "digest-1", "viewer", "digest-2");
+        final MarkRolesAsSyncedTask task = new MarkRolesAsSyncedTask(
+            ActionListener.noop(),
+            TestRestrictedIndices.INTERNAL_SECURITY_MAIN_INDEX_7,
+            null,
+            newRoleDigests
+        );
+        expectThrows(IndexClosedException.class, () -> task.execute(clusterState));
+    }
+
+    public void testMarkRolesAsSyncedTaskRejectsClosingInProgressIndex() {
+        // INDEX_CLOSED_BLOCK is added before IndexMetadata.state flips to CLOSE; the task must reject
+        // both, not just state == CLOSE.
+        final ClusterBlocks closingInProgressBlocks = ClusterBlocks.builder()
+            .addIndexBlock(
+                ProjectId.DEFAULT,
+                TestRestrictedIndices.INTERNAL_SECURITY_MAIN_INDEX_7,
+                MetadataIndexStateService.INDEX_CLOSED_BLOCK
+            )
+            .build();
+        final ClusterState clusterState = markShardsAvailable(createClusterStateWithOpenSecurityIndex()).nodes(localNodeMaster())
+            .blocks(closingInProgressBlocks)
             .build();
         final Map<String, String> newRoleDigests = Map.of("superuser", "digest-1", "viewer", "digest-2");
         final MarkRolesAsSyncedTask task = new MarkRolesAsSyncedTask(

